IN-532 Allow moderators to view reporting

IvaKop opened the pull request:
- IN-532
- [Epic Deployment](https://in532.epic.citizenlab.co/) does not work for some reason, but "it works on my machine".

## What changes are in this PR?

It allows moderators to access insights/reporting tab in the admin panel. They can perform the same operations as admins, but only for their projects.

It could use some additional acceptance tests about moderators, but it was a bit difficult to draw a line where to stop. Permissions in the insights/reporting tab are not granular. If a user has access to a view, the user is allowed to take any action within that view. So, in the end, I just added acceptance tests to `acceptance/views_specs.rb`. Adding acceptance tests for all the other insights resources would have introduce a lot of code just to test the same policies.

<img width="959" alt="image" src="https://user-images.githubusercontent.com/209268/133623995-a4d44c97-5021-42ad-8b1c-41044c661e65.png">

IvaKop committed:
Add insights to moderator exceptions

SaraBCL reviewed (commented):
Perfect ! If you don't mind I'll take over this PR and tsk and add the back-end part to it directly ?

IvaKop commented:
By all means, take over. It makes no sense to merge it without the backend part, anyway.

adessy commented:
Do you mind if I take over the take over? 🙂

adessy committed:
IN-532 Give access to insights to moderators

IN-532 #time 0w 0d 0h 24m Give access to insights to moderators

adessy committed:
IN-532 Adapt the category policy bc the view permissions DO depend on the view now

IN-532 #time 0w 0d 0h 8m Adapt the category policy bc the view permissions DO depend on the view now

adessy committed:
Merge branch 'master' into IN-532/moderator-permission

IN-532 #time 0w 0d 0h 9m Merge branch 'master' into IN-532/moderator-permission

adessy committed:
IN-532 Test view policy + a few fixes

IN-532 #time 0w 0d 1h 1m Test view policy + a few fixes

adessy committed:
IN-532 Adapt view scope policy for moderators

IN-532 #time 0w 0d 0h 36m Adapt view scope policy for moderators

adessy committed:
IN-532 Add acceptance tests for views when user is moderator

IN-532 #time 0w 0d 0h 18m Add acceptance tests for views when user is moderator

adessy committed:
IN-532 Reworking before/after each hooks to be able to use before(:all) in tests

IN-532 #comment before(:all) (and thus let_it_be) are run before(:each). To be able to use them easily, we need to switch to the test tenant before, they are run.

IN-532 #time 0w 0d 3h 11m Reworking before/after each hooks to be able to use before(:all) in tests

adessy committed:
IN-532 Test policy view scope

adessy committed:
IN-532 Remove the category policy which was an overkill as all methods were delegated to the view policy

IN-532 #time 0w 0d 0h 43m Remove the category policy which was an overkill as all methods were delegated to the view policy

adessy committed:
IN-532 Remove unnecessary after(:each) example hook that switch the DB back to the test tenant

IN-532 #comment If necessary, the tenant is switched back as part of the transaction associated to the example (see config.use_transactional_fixtures = true).

More importantly, the db request in the hook was also causing PG::InFailedSqlTransaction errors for some tests that contains (expected) errors in DB queries.

IN-532 #time 0w 0d 2h 42m Remove unnecessary after(:each) example hook that switch the DB back to the test tenant

adessy committed:
Revert "IN-532 Remove unnecessary after(:each) example hook that switch the DB back to the test tenant"

IN-532 #comment This reverts commit e3ba36d3

IN-532 #time 0w 0d 0h 13m Revert "IN-532 Remove unnecessary after(:each) example hook that switch the DB back to the test tenant"

adessy committed:
Revert "IN-532 Reworking before/after each hooks to be able to use before(:all) in tests"

IN-532 #comment This reverts commit 70001bca

IN-532 #time 0w 0d 0h 5m Revert "IN-532 Reworking before/after each hooks to be able to use before(:all) in tests"

adessy committed:
IN-532 Remove let_it_be helpers

IN-532 #comment we need some more refactoring of the test setup to be able to use them

IN-532 #time 0w 0d 0h 3m Remove let_it_be helpers

adessy committed:
Merge branch 'master' into IN-532/moderator-permission

adessy committed:
IN-532 Update CHANGELOG.md

IN-532 #time 0w 0d 0h 51m Update CHANGELOG.md

adessy committed:
[IN-532] Remove unused import

adessy committed:
Merge remote-tracking branch 'origin/IN-532/moderator-permission' into IN-532/moderator-permission

IN-532 #comment # Conflicts:
#	CHANGELOG.md

IN-532 #time 0w 0d 0h 9m Merge remote-tracking branch 'origin/IN-532/moderator-permission' into IN-532/moderator-permission

jinjagit commented on back/engines/commercial/insights/spec/policies/insights/view_policy_spec.rb:
```diff
@@ -0,0 +1,61 @@
+    it { expect(scope.resolve.count).to eq(2) }
+  end
+
+  context "when user moderates the view's project" do
```
Suggestion: `"when user is moderator"` - for consistency, though maybe there is a specific point being made here that I don't get.

jinjagit replied:
I note from the PR description that moderators can _"... perform the same operations as admins, but only for their projects."_, so maybe a better/alternative title is; `"when user moderates their own project(s)"`?

adessy replied:
Yep, today each view is associated to a project and a user can only see the view if (s)he moderates the associated project. What do you think of: *when user moderates the project associated to the view*, or is it even worse? 😄

jinjagit replied:
when user moderates the project associated _with_ the view 👍

jinjagit replied:
but they're all fine tbh - I'm just nitpicking.

jinjagit commented on back/engines/commercial/insights/app/controllers/insights/web_api/v1/categories_controller.rb:
```diff
@@ -34,27 +39,23 @@ def destroy
       def view
-        View.includes(:categories).find(params.require(:view_id))
+        @view ||= authorize(
+          View.includes(:categories).find(params.require(:view_id)),
```
This is the hardest change for me to fully understand.  

My best guess: If `@view` is nil, then authorise and return the `View` that includes the given categories _and_ a specific `view_id`?, and only _if_ it updates successfully?

adessy replied:
Haaa, I see. There are a few things going on here.

First, 
```ruby
def view
	@view ||= ...
end
```
is a pervasive idiom/pattern in ruby that performs memoization. You can see it as a lazy attribute. The attribute is only computed the first time it's used.

The second thing is that the categories controller delegates all the *permissioning* to the view policy (all categories are associated *with* a view). So here is what it's doing:
- it `find`s the right view (`params.require(:view_id)`)
- it already includes the categories along the way to avoid an extra db query when we'll access `view.categories`
- it checks that the current user is allowed to update the said view (`authorize(view, :update?)`)

jinjagit replied:
Thanks. Super-useful explanation! (saving for future reference) 🧑‍🎓

jinjagit reviewed (approved):
All looks good to me. Just one suggestion, and one question (where I don't really understand, but it's probably fine)

adessy committed:
IN-532 Improve the name of the context
